ls1intum/Artemis
# Improve BehaviourTest in java template [extern/old]

### Motivation and Context
BehaviorTest class in templates for Java exercises had bugs and did not provide required functionality for one of the exercises.

### Description
First of all, there was a bug in the method `getParameterTypes` where it would not construct the correct array of classes if any of the parameters were equal in the Java-object sense.
Secondly, there was no way of creating a `newInstance` if the required classes had constructor with primitively-typed arguments. Method `getConstructor` and another overloading of `newInstance` were added to tackle that problem.
Next, there was no way of using `invokeMethod`, but rethrowing an exception which could occur during the underlying method execution. It is sometimes required to check whether implementation of a method by students throws an expected exception. A new method `invokeMethodRethrowing` was added to support this.
Lastly, there are a couple of small refactoring here and there to make the code more readable, and less error-prone.

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a new Java exercise
4. Check the created tests repo, which should include the new version of BehaviorTest class
